File: nodejs/code-duplication/good/orderProcessor.js
const { TaxCalculator, ShippingCalculator } = require('./taxCalculator');

class OrderProcessor {
    constructor() {
        this.taxCalculator = new TaxCalculator();
        this.shippingCalculator = new ShippingCalculator();
    }

    calculateTax(price, state) {
        return this.taxCalculator.calculateTax(price, state);
    }

    calculateShipping(weight, distance) {
        return this.shippingCalculator.calculateShipping(weight, distance);
    }
}

class InvoiceGenerator {
    constructor() {
        this.taxCalculator = new TaxCalculator();
        this.shippingCalculator = new ShippingCalculator();
    }

    calculateTax(price, state) {
        return this.taxCalculator.calculateTax(price, state);
    }

    calculateShipping(weight, distance) {
        return this.shippingCalculator.calculateShipping(weight, distance);
    }
}

class QuoteGenerator {
    constructor() {
        this.taxCalculator = new TaxCalculator();
        this.shippingCalculator = new ShippingCalculator();
    }

    calculateTax(price, state) {
        return this.taxCalculator.calculateTax(price, state);
    }

    calculateShipping(weight, distance) {
        return this.shippingCalculator.calculateShipping(weight, distance);
    }
}

module.exports = {
    OrderProcessor,
    InvoiceGenerator,
    QuoteGenerator
};

File: java/code-duplication/good/OrderProcessor.java
class OrderProcessor {
    private TaxCalculator taxCalculator;
    private ShippingCalculator shippingCalculator;

    public OrderProcessor() {
        this.taxCalculator = new TaxCalculator();
        this.shippingCalculator = new ShippingCalculator();
    }

    public double calculateTax(double price, String state) {
        return taxCalculator.calculateTax(price, state);
    }

    public double calculateShipping(double weight, double distance) {
        return shippingCalculator.calculateShipping(weight, distance);
    }
}

class InvoiceGenerator {
    private TaxCalculator taxCalculator;
    private ShippingCalculator shippingCalculator;

    public InvoiceGenerator() {
        this.taxCalculator = new TaxCalculator();
        this.shippingCalculator = new ShippingCalculator();
    }

    public double calculateTax(double price, String state) {
        return taxCalculator.calculateTax(price, state);
    }

    public double calculateShipping(double weight, double distance) {
        return shippingCalculator.calculateShipping(weight, distance);
    }
}

class QuoteGenerator {
    private TaxCalculator taxCalculator;
    private ShippingCalculator shippingCalculator;

    public QuoteGenerator() {
        this.taxCalculator = new TaxCalculator();
        this.shippingCalculator = new ShippingCalculator();
    }

    public double calculateTax(double price, String state) {
        return taxCalculator.calculateTax(price, state);
    }

    public double calculateShipping(double weight, double distance) {
        return shippingCalculator.calculateShipping(weight, distance);
    }
}
